Data/Primitive/Types/PrimUnlifted.hs

```diff
@@ -0,0 +1,274 @@
+{-# language CPP #-}
+{-# language MagicHash #-}
+{-# language TypeFamilies #-}
+{-# language DataKinds #-}
+{-# language DefaultSignatures #-}
+
+module Data.Primitive.Types.PrimUnlifted
+ ( PrimUnlifted (..)
+#if __GLASGOW_HASKELL__ < 800
+ , Unlift
+ , toArrayArray#
+ , fromArrayArray#
+#endif
+ ) where
+
+import GHC.Exts
+
+import Data.Primitive.PrimArray (PrimArray (..), MutablePrimArray (..))
+import Data.Primitive.Array (Array (..), MutableArray (..))
+import Data.Primitive.SmallArray (SmallArray (..), SmallMutableArray (..))
+import Data.Primitive.ByteArray (ByteArray (..), MutableByteArray (..))
+import Data.Primitive.MutVar (MutVar (..))
+import GHC.MVar (MVar(..))
+import GHC.Conc (TVar(..))
+import GHC.Weak (Weak(..))
+import GHC.Conc.Sync (ThreadId(..))
+
+-- | Classifies the types that are just liftings of unlifted pointer
+-- types.
+--
+-- For GHC versions 8.0 and later, instances should define 'Unlift',
+-- 'toUnlifted#', and 'fromUnlifted#'. When compiled with earlier versions,
+-- @Unlift a@ is always a synonym for 'ArrayArray#', and instances should
+-- define 'toArrayArray#' and 'fromArrayArray#', coercing the internal
+-- unlifted types to and from 'ArrayArray#'.
+--
+-- Note: 'toArrayArray#' and 'fromArrayArray#' will eventually be
+-- deprecated and removed. They should be used /only/ to define instances
+-- for GHC versions before 8.0.
+class PrimUnlifted a where
+#if __GLASGOW_HASKELL__ >= 800
+  {-# MINIMAL (toArrayArray# | toUnlifted#), (fromArrayArray# | fromUnlifted# ) #-}
```

Comment:
Should we change this to `{-# MINIMAL toUnlifted#, fromUnlifted# #-}`? That will produce a technically spurious warning, but we really do want users to get used to defining these methods for 8.0 and up.

Reply:
I agree that `{-# MINIMAL toUnlifted#, fromUnlifted# #-}` is better.